### PR TITLE
Issue #11230: fix FallThroughCheck for labeled statements

### PR DESCRIPTION
**Pull request by relentless-pursuit**
Resolves: #11230

This PR aims to fix FallThroughCheck for labeled statements. 

Diff regression config: https://gist.githubusercontent.com/relentless-pursuit/5f484c60cf4933dcc3cadc564e7e574f/raw/70d35777b174a9d980d09644bb45d1c4fc483365/config_fallThroughCheck.xml

### CODE REVIEW

**Comment by nrmancuso** [comment]
@relentless-pursuit ping, do you plan to continue this PR?

**Comment by relentless-pursuit** [comment]
> @relentless-pursuit ping, do you plan to continue this PR?

Yes, @nrmancuso. Will resume from 1st of Feb. Was caught up with something.

**Comment by nrmancuso** [comment]
> @nrmancuso, in general i struggle with jacoco coverage. And, especially with this, i am struggling to create a jacoco coverage for the while loop i.e. `while (parent !=null)`. Can you provide me some guidance on this?

If we can’t hit that code, just remove it, let CI run, then generate check regression reports.

**Comment by romani** [comment]
@relentless-pursuit , please don't open new PRs, let's finish hanging PRs like this

**Comment by relentless-pursuit** [comment]
> @relentless-pursuit , please don't open new PRs, let's finish hanging PRs like this

Sure.

**Comment by relentless-pursuit** [comment]
@romani, @nrmancuso can we proceed with the review?

**Comment by relentless-pursuit** [comment]
I have used the example mentioned in the test case. Also, tried to replicate the issue with jar with code changes. I didn't see any error. Please let me know if something else needs to be done.

**Review by nrmancuso — CHANGES_REQUESTED** [review]
@relentless-pursuit please do:
1. Generate regression reports
2. Extend test inputs to match https://github.com/checkstyle/checkstyle/pull/12820/files
3. Explain why the implementation has deviated significantly from https://github.com/checkstyle/checkstyle/pull/12820/files#r1141171046

**Comment by relentless-pursuit** [comment]
> 3\. Explain why the implementation has deviated significantly from https://github.com/checkstyle/checkstyle/pull/12820/files#r1141171046

I was focussed on solving the issue based on the input test file. But, after i carefully read and re-read the PR conversation, i realised it is a bit complicated. I was just testing my code for test input mentioned in the issue. Now, I have been able to more or less understand what is aimed here. But, i have a concern. The test input as a part of this PR: https://github.com/checkstyle/checkstyle/pull/12820/files doesn't any longer fit to the recent code base of checksyle as we have changed the method `fallThrough` and done away with `matchesComment`. Hence, when i replicated the PR, and tested with recent Checkstyle master, it showed 2 Fall though violations. 

Functions like `matchesComment` to identify comments based on regular expressions and specific line numbers within the source code, which no longer exist. To my understanding the old PR was almost complete but maybe faced issue on these lines. It couldn't keep up with the changes made so far.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLabeledBreak.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                    }
+                case 10:  // violation 'Fall through from previous branch of the switch statement.'
+                    inner1:
+                    {
+                        switch (a) {
+                            case 1:
+                                break outer; // ok
+                        }
+                    }
+                case 11:
```
Comment (relentless-pursuit) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLabeledBreak.java:95]: If we observe, in the ```
Violation lines for /Users/piyushkumarsadangi/Development/checkstyle/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLabeledBreak.java differ.
unexpected (2): 37, 95
---
expected      : [50, 54, 68, 77, 87]
but was       : [37, 50, 54, 68, 77, 87, 95]

```

Now, ideally we shouldn't be getting any violation at this line: `case 11:`. As we exit the `outer:switch (a)`. However, we continue to get a violation now, which must not have been the case when the old PR was being authored, because the code base has undergone some changes. Now, the question remains do we revert back to old changes, or have a fusion of both worlds, line-number based approach and AST. I am not sure on this. Please let me know if i have understood right. I will continue to understand the existing code, and maybe a change here and there can still get us no violation at 37 and 95. I hope I didn't complicate with my observations.

**Comment by nrmancuso** [comment]
@relentless-pursuit thanks a lot for the explanation; I had completely forgotten about the overhaul to this check. I hope that it makes your life easier in this PR :) Let's generate a regression report while I begin review.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLabeledBreak.java (hunk 1 above)
Comment (nrmancuso) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLabeledBreak.java:95]: Yes, 37 and 95 should not be violations.

**Comment by relentless-pursuit** [comment]
@nrmancuso, I have been debugging for quite some time to see why 37 and 95 are false-positive. And, what i figured out is that they don't go inside `case TokenTypes.SLIST:` inside the method `isTerminated`. Hence, we don't get to execute the `hasOuterLabels` method leading to false positive. 

Now, my understanding of why it goes inside the `case TokenTypes.SLIST:` and `case TokenTypes.LITERAL_BREAK:` is 
when visitToken encounters a switch statement, it examines each case group. The SLIST under a case group includes all the statements related to that case. SLIST acts as the parent node for a sequence of statements. When a `break` statement (`LITERAL_BREAK`) appears directly within the `SLIST` of a case, it is considered a direct child of that `SLIST`. In such scenarios, the `break` statement successfully prevents fall-through. 

Example:
```
case 1:
    switch (b) {
        case 10:
            break top; // ok
        default:
            return 11;
    }


But, if a `break` statment is inside another strcuture like `for` loop in the example, the `for` loop itself becomes the direct child of the `SLIST` associated with the case. The `break` statement is a child of the for loop structure, not the `SLIST` directly. Hence, not directly processed under `LITERAL_BREAK` by the switch-case logic because it's considered part of the for loop's structure, which is inside the `SLIST` like in below example. 

```
        outer: outer2: switch (a) {
            case 1:
                inner1: inner2: switch (b) {
                    case 10:
                        break outer; // ok
                    default:
                        return 11;
                }
            case 2:
                inner: for (int i = 0;i < 3;i++) {
                    break outer2; // ok
                }
            case 3:
                return 2; **// We experience here false-positive**
        }
```

To summarise, `break` statement if directly within the case's statement list (not nested in other structures) it's processed as a `LITERAL_BREAK` in the AST.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java [hunk 2]
```diff
@@ -182,61 +186,83 @@ private boolean isTerminated(final DetailAST ast, boolean useBreak,
...
+
+    /**
+     * Checks if given break or continue ast has outer label.
+     *
+     * @param ast break or continue node
+     * @param labels label names in scope of case group
+     * @return true if local label used
+     */
+    private static boolean hasOuterLabel(DetailAST ast, Set<String> labels) {
+        return ast.getChildCount() == 2 && !labels.contains(ast.getFirstChild().getText());
```
[the remark this answers is not in the document]
Comment (relentless-pursuit) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java:243]: > Please extract `ast.getChildCount() == 2` to some meaningfully named variable so that the next person to code in here knows why this matters.

Specifically, on observing for the below piece fo code which leads to Fall-through false positive, the child count is 3. 
```
 case 2:
                inner: for (int i = 0;i < 3;i++) {
                    break outer2; // ok
                }
```

Some logs from the console:
```
Child 0: Type = 86, Text = break
Child 1: Type = 144, Text = //
Child 2: Type = 73, Text = }
```

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java [hunk 3]
```diff
@@ -246,17 +272,18 @@ private boolean checkSlist(final DetailAST slistAst, boolean useBreak,
...
+                            boolean useContinue, Set<String> labels) {
         final DetailAST thenStmt = getNextNonCommentAst(ast.findFirstToken(TokenTypes.RPAREN));
 
         final DetailAST elseStmt = getNextNonCommentAst(thenStmt);
 
         return elseStmt != null
-                && isTerminated(thenStmt, useBreak, useContinue)
-                && isTerminated(elseStmt.getLastChild(), useBreak, useContinue);
+                && isTerminated(thenStmt, useBreak, useContinue, labels)
+                && isTerminated(elseStmt.getFirstChild(), useBreak, useContinue, labels);
```
[the remark this answers is not in the document]
Comment (relentless-pursuit) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java:286]: The the author of the old PR and some review comments that followed in the old PR agreed on tracing labels for 
`continue` statement as well. Hence, the scope of the PR is not just limited to break but `continue` aswell. 

Reference: https://github.com/checkstyle/checkstyle/pull/12820#discussion_r1141009389

If you find it not required for this PR, i will remove it.

**Comment by relentless-pursuit** [comment]
> @relentless-pursuit thanks a lot for the explanation; I had completely forgotten about the overhaul to this check. I hope that it makes your life easier in this PR :) Let's generate a regression report while I begin review.

I feel the existing code is not sufficient. We still have some work to be done. And, that might break something from past. But, I will proceed with regression report generation.

**Comment by nrmancuso** [comment]
@relentless-pursuit how's it going here? Can we help with something?

**Comment by relentless-pursuit** [comment]
> @relentless-pursuit how's it going here? Can we help with something?

@nrmancuso Sorry for the delay. I have generated a regression report. I did it after long time. let me know if i have done it right, or if the report looks fine. Meanwhile, I had posted some concerns here: https://github.com/checkstyle/checkstyle/pull/14199#issuecomment-1971802298

Please let me know your feedback on the same.

**Comment by nrmancuso** [comment]
> @nrmancuso sorry for the ping. Any feedback on the development so far on this PR?

No need to be sorry :) Feedback from me is that first we need to achieve expected check behavior in new test cases, with no failures in CI. Then we need to generate a regression report. If this is clean, we proceed with review and code refinement. If not, we found new test cases to add and work on. Then rinse, repeat :)

**Comment by relentless-pursuit** [comment]
> > @nrmancuso sorry for the ping. Any feedback on the development so far on this PR?
> 
> No need to be sorry :) Feedback from me is that first we need to achieve expected check behavior in new test cases, with no failures in CI. Then we need to generate a regression report. If this looks good, we proceed with review and code refinement. If not, we found new test cases to add and work on. Then rinse, repeat :)

@nr

> > @nrmancuso sorry for the ping. Any feedback on the development so far on this PR?
> 
> No need to be sorry :) Feedback from me is that first we need to achieve expected check behavior in new test cases, with no failures in CI. Then we need to generate a regression report. If this looks good, we proceed with review and code refinement. If not, we found new test cases to add and work on. Then rinse, repeat :

> @nrmancuso, I have been debugging for quite some time to see why 37 and 95 are false-positive. And, what i figured out is that they don't go inside `case TokenTypes.SLIST:` inside the method `isTerminated`. Hence, we don't get to execute the `hasOuterLabels` method leading to false positive.
> 
> Now, my understanding of why it goes inside the `case TokenTypes.SLIST:` and **NOT** `case TokenTypes.LITERAL_BREAK:` is when visitToken encounters a switch statement, it examines each case group. The SLIST under a case group includes all the statements related to that case. `SLIST` acts as the parent node for a sequence of statements.
> 
> When a `break` statement (`LITERAL_BREAK`) appears directly within the `SLIST` of a case, it is considered a direct child of that `SLIST`. In such scenarios, the `break` statement successfully prevents fall-through.
> 
> ```
> case 1:
>     switch (b) {
>         case 10:
>             break top; // ok
>         default:
>             return 11;
>     }
> ```
> 
> But, if a `break` statment is inside another strcuture like `for` loop in the example, the `for` loop itself becomes the direct child of the `SLIST` associated with the case. The `break` statement is a child of the `for` loop structure, not the `SLIST` directly. Hence, not directly processed under `LITERAL_BREAK` by the switch-case logic because it's considered part of the for loop's structure, which is inside the `SLIST` like in below example.
> 
> ```
>         outer: outer2: switch (a) {
>             case 1:
>                 inner1: inner2: switch (b) {
>                     case 10:
>                         break outer; // ok
>                     default:
>                         return 11;
>                 }
>             case 2:
>                 inner: for (int i = 0;i < 3;i++) {
>                     break outer2; // ok
>                 }
>             case 3: // We experience here false-positive by encountering violation of Fall through
>                 return 2; 
>         }
> ```
> 
> To summarise, `break` statement if directly within the case's statement list (not nested in other structures) it's processed as a `LITERAL_BREAK` in the Checkstyle's existing code.

Kindly ignore. After careful debugging, I realised that `37` and `95` were given false-positive because of not being able to realise break statement because of the // ok comment. Hence, altered the input test file by adding the `// ok` comment to as `// ok below as ...`

**Comment by relentless-pursuit** [comment]
> > @nrmancuso sorry for the ping. Any feedback on the development so far on this PR?
> 
> No need to be sorry :) Feedback from me is that first we need to achieve expected check behavior in new test cases, with no failures in CI. Then we need to generate a regression report. If this looks good, we proceed with review and code refinement. If not, we found new test cases to add and work on. Then rinse, repeat :)

@nrmancuso, updated the regression report. I think the CI failure is unrelated. Please let me know if it is good enough to begin review.

**Comment by relentless-pursuit** [comment]
@nrmancuso ping

**Comment by nrmancuso** [comment]
> @nrmancuso ping

Need a bit of time to get to this one, thanks for the reminder

**Comment by relentless-pursuit** [comment]
> > @nrmancuso ping
> 
> Need a bit of time to get to this one, thanks for the reminder

Sure, will wait.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheckTest.java [hunk 4]
```diff
@@ -365,4 +365,20 @@ public void testReliefCommentBetweenMultipleComment() throws Exception {
                 getPath("InputFallThrough8.java"),
                 expected);
     }
+
+    @Test
```
Comment (nrmancuso) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheckTest.java:369]: Please add another test case with similar inputs, except with switch labeled rules. See https://docs.oracle.com/javase/specs/jls/se21/html/jls-14.html#jls-SwitchRule

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Looking good, please resolve/mark outdated all items (in conversation) from you that have been addressed

Items:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java [hunk 5]
```diff
@@ -182,48 +185,65 @@ private boolean isTerminated(final DetailAST ast, boolean useBreak,
...
 
+    /**
+     * Checks if given break or continue ast has outer label.
+     *
+     * @param ast break or continue node
+     * @param labels label names in scope of case group
+     * @return true if local label used
+     */
+    private static boolean hasOuterLabel(DetailAST ast, Set<String> labels) {
+        final boolean childCount = ast.getChildCount() == 2;
```
[the remark this answers is not in the document]
Comment (relentless-pursuit) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java:231]: > Why do we have this condition?

The name could be deceptive. But, this method helps us to avoid fall-through violation. 

```
        top: switch (a) {
            case 1:
                switch (b) {
                    case 10:
                        break top; // ok as terminates outer 'top' switch
                    default:
                        return 11;
                }
            case 2:
                return 2;
        }
```

In the above scenario,` hasOuterLabel` helps us to come out of the entire scope of `top`. And, this prevents fall-through as we don't fall onto case 2 after `break top; `

```
        top: switch (a) {
            case 1:
                inner: switch (b) {
                    case 10:
                        break inner;                     
                        return 11;
                }
            case 2: // Fall through
                return 2;
        }
```

Now, here, we have again a labeled break. We break out of scope of inner switch block and control comes to case 2 causing Fall throgh violation. In nutshell, this method only helps in detecting a labeled break, be it inner or outer.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheckTest.java (hunk 4 above)
Comment (relentless-pursuit) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheckTest.java:369]: > Please add another test case with similar inputs, except with switch labeled rules. See https://docs.oracle.com/javase/specs/jls/se21/html/jls-14.html#jls-SwitchRule

I am a bit confused. I am reading the documentation. And, it mentions of SwitchRule and SwitchLabel. 

```
int method(int a, int b) {
    top:
    switch (a) {
        case 1 -> {
            switch (b) {
                case 10 -> { break top; } // ok as terminates outer 'top' switch
                default -> { return 11; }
            }
        }
        case 2 -> { return 2; }
    }
}

```

Something like this? Like replacing with `->` ?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java:231]: fwiw, I doubt that your second example would compile; we will never reach the `return 11;` statement.


> The name could be deceptive.

Yes, it is. This is not good. 

>childCount

This sounds like it should be a numerical value.


Please do something like this to make it very clear what is happening here:
```suggestion

    private static boolean hasOuterLabel(DetailAST statement, Set<String> labels) {
        final boolean isLabeledStatement = ast.getChildCount() == 2;
        final String identifier = ast.getFirstChild().getText
        return childCount && !labels.contains(identifier);
    }
````

Or to be even more clear and avoid possibility of  `;` being called an identifier or a possible NPE, we could do something like this:
```suggestion

    private static boolean hasOuterLabel(DetailAST statement, Set<String> labels) {       
        return Optional.of(statement)
            .map(DetailAST::getFirstChild)  // get first child
            .filter(child -> child.getType() == TokenTypes.IDENT) // see if it is an identifier
            .map(DetailAST::getText) // if so, get the name
            .filter(label -> !labels.contains(label)) // see if this label is in labels
            .isPresent();
    }

```

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java [hunk 6]
```diff
@@ -169,10 +171,11 @@ public void visitToken(DetailAST ast) {
      * @param ast root of given subtree
      * @param useBreak should we consider break as terminator
      * @param useContinue should we consider continue as terminator
+     * @param labels label names
      * @return true if the subtree is terminated.
      */
     private boolean isTerminated(final DetailAST ast, boolean useBreak,
-                                 boolean useContinue) {
+                                 boolean useContinue, Set<String> labels) {
```
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java:178]: Each Set of labels is only for a given "level" of switch, right?

Example of my understanding:
```
        top: switch (a) { // 'top' is in one Set of labels
            case 1:
                inner: switch (b) { // 'inner' is in a different Set of labels
                    case 10:
                        break inner;                     
                }
            case 2: // Fall through
                return 2;
        }
```

If my understanding is correct, can we somehow make this obvious? Perhaps improve the Javadoc, and rename this Set to something that makes this obvious?

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Item:

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheckTest.java (hunk 4 above)
Comment (nrmancuso) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheckTest.java:369]: > Like replacing with -> ?

Basically, yes. There are some different AST elements around this, so it will be good to make sure we are covered there.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (relentless-pursuit) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java:231]: > fwiw,

Yes, it overlooked it. but, the idea was to explain what haOuterLabel does. it is not a part of the PR.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheckTest.java (hunk 4 above)
Comment (nrmancuso) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheckTest.java:369]: Just wanted to be clear here, there is no possibility of fallthrough with this syntax, I just want to make sure there is no possibility of NPE.

Comment (relentless-pursuit) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheckTest.java:369]: > Basically, yes. There are some different AST elements around this, so it will be good to make sure we are covered there.

Have added a test-case and corresponding input file using switch rules. Let me know your thoughts on the same.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (relentless-pursuit) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java]: > fwiw, I doubt that your second example would compile; we will never reach the `return 11;` statement.
> 
> > The name could be deceptive.
> 
> Yes, it is. This is not good.
> 
> > childCount
> 
> This sounds like it should be a numerical value.
> 
> Please do something like this to make it very clear what is happening here:
> 
> Or to be even more clear and avoid possibility of `;` being called an identifier or a possible NPE, we could do something like this:

Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java (hunk 6 above)
Comment (relentless-pursuit) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java]: > Each Set of labels is only for a given "level" of switch, right?
> 
> Example of my understanding:
> 
> ```
>         top: switch (a) { // 'top' is in one Set of labels
>             case 1:
>                 inner: switch (b) { // 'inner' is in a different Set of labels
>                     case 10:
>                         break inner;                     
>                 }
>             case 2: // Fall through
>                 return 2;
>         }
> ```
> 
> If my understanding is correct, can we somehow make this obvious? Perhaps improve the Javadoc, and rename this Set to something that makes this obvious?

Yes, you are right. I have made changes to the the name.

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Please make CI happy, we are almost there:

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheckTest.java (hunk 4 above)
Comment (nrmancuso) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheckTest.java:369]: Looks good. Thanks!

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java [hunk 7]
```diff
@@ -182,48 +187,78 @@ private boolean isTerminated(final DetailAST ast, boolean useBreak,
...
     }
 
+    /**
+     * Checks if given break or continue ast has outer label.
+     *
+     * @param ast break or continue node
+     * @param labelsForCurrentSwitchScope the Set labels for the current scope of the switch
+     * @return true if local label used
+     */
+    private static boolean hasLabels(DetailAST ast, Set<String> labelsForCurrentSwitchScope) {
```
[the remark this answers is not in the document]
Comment (relentless-pursuit) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java]: Yes, thanks for pointing it out. Corrected.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java [hunk 8]
```diff
@@ -182,48 +187,78 @@ private boolean isTerminated(final DetailAST ast, boolean useBreak,
...
+        return Optional.ofNullable(ast)
+                .filter(node -> node.getChildCount() == 2)
+                .map(node -> node.getFirstChild())
```
[the remark this answers is not in the document]
Comment (relentless-pursuit) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java]: Done. Waiting for CI's status.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java [hunk 9]
```diff
@@ -182,48 +187,78 @@ private boolean isTerminated(final DetailAST ast, boolean useBreak,
...
+     * @param statement break or continue node
+     * @param labelsForCurrentSwitchScope the Set labels for the current scope of the switch
+     * @return true if local label used
+     */
+    private static boolean hasLabel(DetailAST statement, Set<String> labelsForCurrentSwitchScope) {
+        return Optional.ofNullable(statement)
+                .filter(node -> node.getChildCount() == 2)
+                .map(DetailAST::getFirstChild)
+                .map(DetailAST::getText)
+                .map(text -> !labelsForCurrentSwitchScope.contains(text))
```
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java:246]:
```suggestion
                .map(label -> !labelsForCurrentSwitchScope.contains(label))
```

**Review by nrmancuso — CHANGES_REQUESTED** [review]
One more minor one:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java (hunk 9 above)
Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java:246]: @relentless-pursuit was there an issue with filtering by TokenTypes.IDENT as suggested in https://github.com/checkstyle/checkstyle/pull/14199#discussion_r1525031818?

Comment (relentless-pursuit) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java:246]: @nrmancuso I missed it somehow. However, i see failed cases, and more violations expected than it is now. 

```
    private static boolean hasLabel(DetailAST statement, Set<String> labelsForCurrentSwitchScope) {
        return Optional.ofNullable(statement)
                .filter(child -> child.getType() == TokenTypes.IDENT)
                .map(DetailAST::getFirstChild)
                .map(DetailAST::getText)
                .map(text -> !labelsForCurrentSwitchScope.contains(text))
                .orElse(Boolean.FALSE);
    }
```

```
Violation lines for /Users/piyushkumarsadangi/Development/checkstyle/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLabeledBreak.java differ.
unexpected (7): 22, 33, 38, 64, 74, 83, 98
---
expected      : [51, 55, 69, 78, 88]
but was       : [22, 33, 38, 51, 55, 64, 69, 74, 78, 83, 88, 98]
```

```
    int method(int a, int b) {
        top: switch (a) {
            case 1:
                switch (b) {
                    case 10:
                        break top; // ok as terminates outer 'top' switch
                    default:
                        return 11;
                }
            case 2: // Expecting a Fall-through violation here, which shouldn't be the case
                return 2;
        }
        
        ...
```

Comment (relentless-pursuit) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java]: @nrmancuso , please ignore my above comment. I hurriedly tested without carefully observing your comment. 
Yes, it seems to be working. 

```
    private static boolean hasLabels(DetailAST statement, Set<String> labels) {
        return Optional.of(statement)
                .map(DetailAST::getFirstChild)  // get first child
                .filter(child -> child.getType() == TokenTypes.IDENT) // see if it is an identifier
                .map(DetailAST::getText) // if so, get the name
                .filter(label -> !labels.contains(label)) // see if this label is in labels
                .isPresent();
    }
```

I will push my code changes again with your suggestions. Thanks for your patience.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java [hunk 10]
```diff
@@ -182,48 +187,78 @@ private boolean isTerminated(final DetailAST ast, boolean useBreak,
...
+     * @param labelsForCurrentSwitchScope the Set labels for the current scope of the switch
+     * @return true if local label used
+     */
+    private static boolean hasLabel(DetailAST statement, Set<String> labelsForCurrentSwitchScope) {
+        return Optional.ofNullable(statement)
+                .map(DetailAST::getFirstChild)
+                .filter(child -> child.getType() == TokenTypes.IDENT)
+                .map(DetailAST::getText)
+                .map(label -> !labelsForCurrentSwitchScope.contains(label))
+                .orElse(Boolean.FALSE);
```
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java:247]:
```suggestion
                .ifPresent();
```

is cleaner and should make checker happy

Comment (relentless-pursuit) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java:247]: I actually had isPresent() but the I guess one of the CI job failed, stating some issue. I can't find it. And, had to look for something else. Will try to revert it.

Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java]: I am good to suppress checker violation for this

**Comment by relentless-pursuit** [comment]
> @relentless-pursuit please make CI happy, then I will hand off to the next reviewer. Thanks for this!

@nrmancuso, It tried looking into the logs, but couldn't find why it is failing. Though, i can see 2 warnings for `FallThrougCheck.java` which is related to `Optional` use, one of them being used before this PR. And, I did add the suppression too. Can you guide me where in the logs, it is leading to failure? Is it failing for some threshold level of warnings allowed, and we have a new warning added due to `hasLabel` `Optional` approach.

**Comment by nrmancuso** [comment]
Let’s try to apply patch one more time, it looks like ordering changed

**Comment by relentless-pursuit** [comment]
> Let’s try to apply patch one more time, it looks like ordering changed

I tried that by placing suppressions of `FallThroughCheck.java` together. But, the failure persists.

**Comment by nrmancuso** [comment]
> > Let’s try to apply patch one more time, it looks like ordering changed
> 
> I tried that by placing suppressions of `FallThroughCheck.java` together. But, the failure persists.

you should only `git apply …` the patch that is provided in the failing CI job

**Comment by relentless-pursuit** [comment]
> > > Let’s try to apply patch one more time, it looks like ordering changed
> > 
> > 
> > I tried that by placing suppressions of `FallThroughCheck.java` together. But, the failure persists.
> 
> you should only `git apply …` the patch that is provided in the failing CI job

The failing job is suceeding now. I assumed patch meant something else, i.e. hiding the suppresion. However, I wasn't aware of git apply. I learnt something new. Thank you.

**Comment by relentless-pursuit** [comment]
@nrmancuso ping

**Review by nrmancuso — APPROVED** [review]
Good stuff!

**Comment by relentless-pursuit** [comment]
@romani ping

**Review by romani — APPROVED** [review]
ok to merge
